cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -57,45 +57,70 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr) {
     case substrait::Expression::kSelection: {
       if (!expr.selection().has_direct_reference()) break;
 
-      util::optional<compute::Expression> root_expr;
+      util::optional<compute::Expression> out;
       if (expr.selection().has_expression()) {
-        ARROW_ASSIGN_OR_RAISE(root_expr, FromProto(expr.selection().expression()));
+        ARROW_ASSIGN_OR_RAISE(out, FromProto(expr.selection().expression()));
       }
 
-      const auto& ref = expr.selection().direct_reference();
-      switch (ref.reference_type_case()) {
-        case substrait::Expression::ReferenceSegment::kStructField: {
-          if (ref.struct_field().has_child()) break;
-
-          FieldRef out(ref.struct_field().field());
+      const auto* ref = &expr.selection().direct_reference();
+      while (ref != NULLPTR) {
+        switch (ref->reference_type_case()) {
+          case substrait::Expression::ReferenceSegment::kStructField: {
+            auto index = ref->struct_field().field();
+            if (!out) {
+              // Root StructField (column selection)
+              out = compute::field_ref(FieldRef(index));
+            } else if (auto out_ref = out->field_ref()) {
+              // Nested StructFields on the root (selection of struct-typed column
+              // combined with selecting struct fields)
+              out = compute::field_ref(FieldRef(*out_ref, index));
+            } else if (out->call() && out->call()->function_name == "struct_field") {
+              // Nested StructFields on top of an arbitrary expression
+              std::static_pointer_cast<arrow::compute::StructFieldOptions>(
+                  out->call()->options)
+                  ->indices.push_back(index);
+            } else {
+              // First StructField on top of an arbitrary expression
+              out = compute::call("struct_field", {std::move(*out)},
+                                  arrow::compute::StructFieldOptions({index}));
+            }
```

Comment:
Instead of inspecting the function name, we could declare `arrow::compute::StructFieldOptions* struct_field_options = nullptr;` outside the loop, then:
```suggestion
            } else if (struct_field_options != nullptr) {
              // Nested StructFields on top of an arbitrary expression
              struct_field_options->indices.push_back(index);
            } else {
              // First StructField on top of an arbitrary expression
              out = compute::call("struct_field", {std::move(*out)},
                                  arrow::compute::StructFieldOptions({index}));
              struct_field_options = checked_cast<arrow::compute::StructFieldOptions*>(out->call()->options.get());
            }
```

Reply:
That (on its own) wouldn't catch this case:

```js
{
  "selection": {
    "directReference": {
      "structField": {
        "field": 12
      }
    },
    "expression": {
      "selection": {
        "directReference": {
          "structField": {
            "field": 2
          }
        },
        "expression": { /*...*/ }
      }
    }
  }
}
```

i.e. it would turn it into `struct_field(struct_field(..., {2}), {12})` rather than the `struct_field(..., {2, 12})` that I would personally expect, since `out` would start off as a `struct_field` call in the outermost `FromProto()` while `struct_field_options` is its initial null value. I could of course initialize the pointer based on what's returned by the nested `FromProto()` rather than always to null, but that just seems less readable/more error-prone/code repetition to me, for marginal acceleration of the conversion code.

You might argue that it *should* return `struct_field(struct_field(..., {2}), {12})`, though? This is kind of what I meant when I asked about whether these conversions should "optimize" (canonicalize might have been a better term), output the most trivial scheme that's correct, go out of their way to maintain structure as much as possible, or something in between. I usually gravitate to the former in conversions like this.

Reply:
I think we want to maintain structure for these functions; canonicalizing passes are definitely worthwhile but should be separated from conversion passes- otherwise roundtripping becomes more difficult to navigate